aten/src/ATen/core/type.cpp

```diff
@@ -748,6 +752,104 @@ TupleTypePtr TupleType::createNamed(
       field_types, qualName, schema)); // NOLINT(modernize-make-shared)
 }
 
+void flatten_union(TypePtr& type, std::vector<TypePtr>& res) {
+  if (type->kind() == UnionType::Kind) {
+    for (auto inner_type : type->expect<UnionType>()->types()) {
+      flatten_union(inner_type, res);
+    }
+  }
+  else if (type->kind() == OptionalType::Kind) {
+    auto inner_type = type->expect<OptionalType>()->getElementType();
+    if (inner_type->kind() == UnionType::Kind) {
+      flatten_union(inner_type, res);
+    }
+    else {
+      res.emplace_back(inner_type);
+    }
+    res.emplace_back(NoneType::get());
+  }
+  else {
+    res.emplace_back(type);
+  }
+}
+
+UnionType::UnionType(std::vector<TypePtr> types) :
+  Type(TypeKind::UnionType) {
+  // Flatten Unions of Unions while populating `types_`
+  for (auto type : types) {
+    flatten_union(type, types_);
+  }
+  // We want the elements to be sorted so we can easily compare two UnionType
+  // objects for equality in the future
+  std::sort(types_.begin(), types_.end(),
```

Comment:
this will only guarantee stable sorting for types with the same kind, but e.g Union[List[int], List[float]] types will not equal Union[List[float], List[int]]. I think we're relying on the ordering in `==` and maybe elsewhere